repair/row_level.cc

```
@@ -1090,69 +1090,58 @@ class repair_meta {
+        for (auto it = _row_buf.rbegin(); it != _row_buf.rend(); ++it) {
             // Move the rows > _current_sync_boundary to _working_row_buf
             // Delete the rows > _current_sync_boundary from _row_buf
             // Swap _working_row_buf and _row_buf so that _working_row_buf
             // contains rows within (_last_sync_boundary,
             // _current_sync_boundary], _row_buf contains rows wthin
             // (_current_sync_boundary, ...]
-            return repeat([this, &it] () {
-                if (it == _row_buf.rend()) {
-                    return make_ready_future<stop_iteration>(stop_iteration::yes);
-                }
-                repair_row& r = *(it++);
-                if (_cmp(r.boundary(), *_current_sync_boundary) > 0) {
-                    _working_row_buf.push_front(std::move(r));
-                    return make_ready_future<stop_iteration>(stop_iteration::no);
-                }
-                return make_ready_future<stop_iteration>(stop_iteration::yes);
-            }).then([this, sz] {
-                _row_buf.resize(_row_buf.size() - _working_row_buf.size());
-                _row_buf.swap(_working_row_buf);
-                if (sz != _working_row_buf.size() + _row_buf.size()) {
-                    throw std::runtime_error(format("incorrect row_buf and working_row_buf size, before={}, after={} + {}",
-                            sz, _working_row_buf.size(), _row_buf.size()));
-                }
-            });
-        });
+            repair_row& r = *it;
+            if (_cmp(r.boundary(), *_current_sync_boundary) <= 0) {
+                break;
+            }
+            _working_row_buf.push_front(std::move(r));
+            co_await coroutine::maybe_yield();
+        }
+        _row_buf.resize(_row_buf.size() - _working_row_buf.size());
+        _row_buf.swap(_working_row_buf);
+        if (sz != _working_row_buf.size() + _row_buf.size()) {
+            co_await coroutine::return_exception(
```

Can't we just call throw here? Why co_await coroutine::return_exception?

---

good point. i thought this was the recommended approach for no reasons. after a second thought, i realized that `coroutine::return_exception()`  was designed for the case where we want to propagate an existing exception. will change to a plain `throw` statement here.